src/react/reconcilation.js

```diff
@@ -1448,4 +1475,10 @@ export class Reconciler {
       }
     }
   }
+
+  _throwIfRenderWasNotPure(effects: Effects, evaluatedNode: ReactEvaluatedNode): void {
+    // TODO: should we also check realm.savedEffects?
```

Comment:
Hopefully effects lists everything the render did. It would be weird and brittle if additional effects were still lurking in realm.savedEffects.